example/cajita_tutorial/07_local_mesh/local_mesh_example.cpp

```diff
@@ -0,0 +1,141 @@
+/****************************************************************************
+ * Copyright (c) 2018-2021 by the Cabana authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <Cajita.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <iostream>
+
+#include <mpi.h>
+
+//---------------------------------------------------------------------------//
+// Local Mesh example.
+//---------------------------------------------------------------------------//
+void localMeshExample()
+{
+    /*
+      Just as the local grid describes the subdomain of the global grid, the
+      local mesh describes a subdomain of the global mesh. In contrast to the
+      global data structures, the local mesh is created using the local grid.
+      This is because the local indexing in required to correctly extract the
+      necessary information from the local mesh.
+    */
+
+    int comm_rank = -1;
+    MPI_Comm_rank( MPI_COMM_WORLD, &comm_rank );
+
+    if ( comm_rank == 0 )
+    {
+        std::cout << "Cajita Local Mesh Example" << std::endl;
+        std::cout << "    (intended to be run with MPI)\n" << std::endl;
+    }
+
+    using exec_space = Kokkos::DefaultHostExecutionSpace;
+    using device_type = exec_space::device_type;
+
+    /*
+      As with the previous examples, we first create the global mesh and grid,
+      as well as the local grid.
+    */
+    std::array<double, 3> low_corner = { -1.2, 0.1, 1.1 };
+    std::array<double, 3> high_corner = { -0.3, 9.5, 2.3 };
+    double cell_size = 0.05;
+    auto global_mesh =
+        Cajita::createUniformGlobalMesh( low_corner, high_corner, cell_size );
+
+    // Here we partition only in x to simplify the example below.
+    int comm_size;
+    MPI_Comm_size( MPI_COMM_WORLD, &comm_size );
+    std::array<int, 3> input_ranks_per_dim = { comm_size, 1, 1 };
+    Cajita::ManualBlockPartitioner<3> partitioner( input_ranks_per_dim );
+
+    // Create the global grid.
+    std::array<bool, 3> is_dim_periodic = { true, true, true };
+    auto global_grid = Cajita::createGlobalGrid( MPI_COMM_WORLD, global_mesh,
+                                                 is_dim_periodic, partitioner );
+
+    // Create a local grid
+    int halo_width = 1;
+    auto local_grid = Cajita::createLocalGrid( global_grid, halo_width );
+
+    // Create the local mesh from the local grid.
```

Comment:
The `device_type` template indicates which memory and execution spaces will be able to access the resulting local mesh data structure.